### ISSUE TITLE
need add feature "ipfs p2p listen /x/test /ipfs/QmXXXXXXXXX/x/test1"

### ISSUE THREAD

**Issue by mccoysc** [ipfs/go-ipfs#5735]
need add feature "ipfs p2p listen /x/test /ipfs/QmXXXXXXXXX/x/test1"

now ipfs p2p listen can only forward data to an ip:port,
maybe,support forward data to another node's protocol "/x/test1" is important.
for eaample:
ipfs p2p listen /x/test /ipfs/QmXXXXXXXXX/x/test
this would make the current node listen at protocol "/x/test" and forward the data to node "/ipfs/QmXXXXXXXXX" and it's protocol "/x/test1"

**Comment by mccoysc** [comment]
supported by these feature of p2p,ipfs can make a transport layer which don't depend on ip:port,just only the "peerId"

**Comment by Stebalien** [comment]
I see, so you want to forward data from one IPFS node to another? That is, you want to forward inbound connections on protocol `/x/one` to `/x/two` on some *other* IPFS node?

That sounds useful but what's your specific use-case? You might better be served with the [relay](https://github.com/ipfs/go-ipfs/blob/master/docs/experimental-features.md#circuit-relay) feature.

**Comment by mccoysc** [comment]
@Stebalien yes,if this feature is done,we can use this feature to make a "named network".
"named network" is a new network layer and has some uniqueness of TCP/IP.
for exampel:security,non-attackable and so on.

**Comment by mccoysc** [comment]
in legacy network,all the security problems are based on the TCP/IP network layer,if IP and port is not used for data transfering,no one can attack your servers or business.

**Comment by mccoysc** [comment]
and also,we need a way to directly push some data to a named network and receive the data from another node ,other than ip:port-node-node-ip:port

**Comment by mccoysc** [comment]
maybe these features are needed:
1,ip:port-node-node-ip:port.
2,ip:port-node-node-node.
3,ip:port-node-node-cmd.
4,node-node-node-ip:port
5,cmd-node-node-ip:port
....

and so on.

**Comment by Stebalien** [comment]
I see. You basically want p2p proxies. Note: you *can* technically do this now with multiple p2p tunnels. That is, you can forward `/x/test -> localhost:1234` and then forward `localhost:1234 -> /x/test on /ipfs/OtherNode`. However, that's not the most efficient way to do this.

**Comment by mccoysc** [comment]
@Stebalien not only "P2P proxy",but a new network layer without tcp/ip.

**Comment by Stebalien** [comment]
So, under the covers we use something called libp2p. You should check it out, I think you'd be *very* interested: https://github.com/libp2p/go-libp2p.

* Examples: https://github.com/libp2p/go-libp2p-examples
* JS Version: https://github.com/libp2p/js-libp2p (yes, it works in web browsers).
* WIP stand alone daemon: https://github.com/libp2p/go-libp2p-daemon/

**Comment by mccoysc** [comment]
@Stebalien yes,it's the extract……but my need is not only the 'code', the 'so many ipfs node' is more important and useful for a network.
or,i would make my own network and enough nodes

**Comment by Stebalien** [comment]
Unless you explicitly configure libp2p not to, libp2p nodes will automatically join the ipfs network. To be precise, the network is actually the *libp2p* network, IPFS nodes are just a subset.
